src/wp-includes/functions.php

```diff
@@ -6039,6 +6039,45 @@ function _doing_it_wrong( $function_name, $message, $version ) {
 	}
 }
 
+/**
+ * Generates a user-level error/warning/notice/deprecation message.
+ *
+ * Generates the message when `WP_DEBUG` is true.
+ *
+ * @since 6.4.0
+ *
+ * @param string $function_name The function that triggered the error.
+ * @param string $message       The message explaining the error.
+ * @param int    $error_level   Optional. The designated error type for this error.
+ *                              Only works with E_USER family of constants. Default E_USER_NOTICE.
+ */
+function wp_trigger_error( $function_name, $message, $error_level = E_USER_NOTICE ) {
+
+	// Bail out if WP_DEBUG is not turned on.
+	if ( ! WP_DEBUG ) {
+		return;
+	}
+
+	/**
+	 * Fires when the given function triggers a user-level error/warning/notice/deprecation message.
+	 *
+	 * Can be used for debug backtracking.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param string $function_name The function that was called.
+	 * @param string $message       A message explaining what has been done incorrectly.
+	 * @param int    $error_level   The designated error type for this error.
+	 */
+	do_action( 'wp_trigger_error_run', $function_name, $message, $error_level );
+
+	if ( ! empty( $function_name ) ) {
+		$message = sprintf( '%s(): %s', $function_name, $message );
+	}
+
+	trigger_error( $message, $error_level );
```

Comment:
This would probably be a bit better if it was using `esc_html()` (if available) for the message.

Reply:
None of the existing `trigger_error()` instances in Core use `esc_html()` including `_doing_it_wrong()` and `_deprecated_*()` functions. Should they be?

The error messages are intended for the server logs. For those messages, escaping isn't necessary.

But how when errors are posted to the screen?

I think the messages are hardcoded, not pulled from the database or user input. Right? If yes, then escaping shouldn't be necessary either.

Reply:
Thinking about this more ... `trigger_error()` is for hardcoded PHP error/warning/notice/deprecation messages. I don't think they need escaping.

Reply:
Passing a `$message` of `'<script>alert( "Howdy!" );</script>'` will output an alert to the browser window. It's theoretically possible that a `wp_trigger_error()` call could include a `$message` parameter containing user input or the like. Escaping late here might be worth doing.

However it's worth noting that the other functions don't currently do this, but would start to if we implement escaping here and make the planned change from their `trigger_error()` calls to `wp_trigger_error()`.

Reply:
Just tested. The output of `trigger_error()` is a `string` that is wrapped in double quotes. Thus any HTML is output as a string, rather than a valid HTML element within the browser.

For example:

```php
wp_trigger_error( '', 'Message <strong>should not include</strong> <script>alert( "Howdy!" );</script> the function as this is not within a function scope.' );
```

outputs:
```html
<br>
<b>Notice</b>:  Message &lt;strong&gt;should not include&lt;/strong&gt; &lt;script&gt;alert( "Howdy!" );&lt;/script&gt; the function as this is not within a function scope. in <b>/var/www/src/wp-includes/functions.php</b> on line <b>6080</b><br>
```

Thus, escaping is not needed.

Reply:
But interesting enough, the [PHP page has a warning callout in the `Notes` section](https://www.php.net/manual/en/function.trigger-error.php):

>**Warning** HTML entities in message are not escaped. Use [htmlentities()](https://www.php.net/manual/en/function.htmlentities.php) on the message if the error is to be displayed in a browser.

Hmm interesting.

Reply:
Oh wait - egg on my face - I tested the above with `esc_html( $message )` 🤦‍♀️  So of course the output is an escaped string. Doh.

Reply:
Appears it does need to be escaped if the message is to be displayed in the browser.

Reply:
Heh, yes, thinking this is the reason why most (all?) other strings/messages that come from PHP or WP, like in `die()` are HTML escaped.

Reply:
Applying this new function in each `_deprecated_*()` and `_doing_it_wrong()` function ensures their messages are also escaped. Cool.

Reply:
Great catch @azaozz ! With that, I think this PR is ready for commit. 

For using it in `_doing_it_wrong()` and `_deprecated_*()` functions, I think I'll do a follow-up PR as the batch likely needs a proper discussion.